sequencer/statetransition.go

```diff
@@ -0,0 +1,169 @@
+
+		for {
+			select {
+			case <-s.ctx.Done():
+				log.Infow("state transition processor stopped")
+				return
+			case <-ticker.C:
+				s.processPendingTransitions()
+			}
+		}
+	}()
+	return nil
+}
+
+func (s *Sequencer) processPendingTransitions() {
+	// copy pids to avoid locking the map for too long
+	s.pidsLock.RLock()
+	pids := make(map[string]time.Time, len(s.pids))
+	maps.Copy(pids, s.pids)
+	s.pidsLock.RUnlock()
+	// iterate over the process IDs and process the ones that are ready
+	for pid := range pids {
+		// check if there is a batch ready for processing
+		batch, batchID, err := s.stg.NextBallotBatch([]byte(pid))
+		if err != nil {
+			if err != storage.ErrNoMoreElements {
+				log.Errorw(err, "failed to get next ballot batch")
+			}
+			continue
+		}
+		// if the batch is nil, skip it
+		if batch == nil || len(batch.Ballots) == 0 {
+			log.Debugw("no ballots in batch", "batchID", batchID)
+			continue
+		}
+		// decode process ID and load metadata
+		processID := new(types.ProcessID).SetBytes(batch.ProcessID)
+		log.Debugw("state transition ready for processing",
+			"processID", processID.String(),
+			"ballotCount", len(batch.Ballots))
+		startTime := time.Now()
+		// process the batch to get the proof
+		proof, err := s.processStateTransitionBatch(processID, batch)
+		if err != nil {
+			log.Errorw(err, "failed to process state transition batch")
+			continue
+		}
+		log.Debugw("state transition proof generated", "took", time.Since(startTime).String())
+		// store the proof in the state transition storage
+		if err := s.stg.PushStateTransitionBatch(&storage.StateTransitionBatch{
+			ProcessID: batch.ProcessID,
+			Proof:     proof.(*groth16_bn254.Proof),
+			Ballots:   batch.Ballots,
+		}); err != nil {
+			log.Errorw(err, "failed to push state transition batch")
+			continue
+		}
+		// mark the batch as done
+		if err := s.stg.MarkBallotBatchDone(batchID); err != nil {
+			log.Errorw(err, "failed to mark ballot batch as done")
+			continue
+		}
+		// update the last update time
+		s.pidsLock.Lock()
+		s.pids[pid] = time.Now()
+		s.pidsLock.Unlock()
+	}
+}
+
+func (s *Sequencer) processStateTransitionBatch(
+	processID *types.ProcessID,
+	batch *storage.AggregatorBallotBatch,
+) (groth16.Proof, error) {
+	// initialize the process state
+	processState, err := s.loadState(processID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create state: %w", err)
+	}
+	// generate the state transition assignments from the batch
+	assignments, err := s.stateBatchToWitness(processState, batch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate assignments: %w", err)
+	}
+	// generate the state transition witness
+	witness, err := frontend.NewWitness(assignments, circuits.StateTransitionCurve.ScalarField())
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate witness: %w", err)
+	}
+	// generate the proof
+	proof, err := groth16.Prove(s.statetransitionCcs, s.statetransitionProvingKey, witness)
```

Comment:
You need to use option `solidity.WithProverTargetSolidityVerifier(backend.GROTH16)` in order to make it compatible with the Solidity verifier.

See: https://github.com/vocdoni/circom2gnark/blob/main/test/recursion_circuit_to_solidity_test.go#L156